NewUI: make card heights consistent

Sudarshan-21 opened the pull request:
`closes` #117 

- [x] Consistent card heights achieved
- [x]  Settle the `open` button to the bottom of card.

**Description**
In this PR I changed  `<footer class="pt-2 pb-3">` to `<footer class="mt-2 mb-1">`. The reason for this particular change is that: padding in the component is not needed. we just need to have some space between different elements which can be achieved by margin(which is a preferred way).

benoit74 commented:
@Sudarshan-21 thank you! Once this PR is ready, please share here a screenshot of the before / after situation so that our PM can decide if this makes sense or not

Sudarshan-21 commented:
> @Sudarshan-21 thank you! Once this PR is ready, please share here a screenshot of the before / after situation so that our PM can decide if this makes sense or not

sure.

benoit74 commented:
@Popolechien do you prefer the UI "before" or "after" this change? 

The idea is that in some **rare** occasions (not encountered so far on real channels, only on test stuff where I merged content from multiple channels into a test channel), the topic image might not be always the same height, and it leads to a weird UI.

@Sudarshan-21 considers that current new UI ("before" screenshot above) is not adequate and is proposing to change it to the "after" UI where all cards have the same height.

Personally I tend to prefer the "before" UI because images are never distorted, while they are in the "after" situation (and cropping instead of distorting them would not help much since it would cut some text at a random position, since we have no control on where important content is placed in the image). I agree it is not perfect, but I have no clue about how to make it perfect without ensuring that images in the channel we use are always the same height for a given topic (which is most probably the case 99% of the time).

I'm quite sure that distorting images will cause more users to complain than having cards of various heights.

WDYT?

Popolechien commented:
> The idea is that in some rare occasions (not encountered so far on real channels

@benoit74 So basically implementing it is not likely to have any material effect on current display (except for rare and yet-unseen-in-the-wild-occasions), correct?

Regarding the distortion: how hard is it to set the default height to that of the tallest image, and for all other images adapt the text region to make up for it accordingly (which means that image height would not be hardcoded but relative to the available set)? 

In either case (distortion or not), having cards at the same height would be my preferred path going forward.

benoit74 commented:
> So basically implementing it is not likely to have any material effect on current display (except for rare and yet-unseen-in-the-wild-occasions), correct?

Correct

> Regarding the distortion: how hard is it to set the default height to that of the tallest image, and for all other images adapt the text region to make up for it accordingly (which means that image height would not be hardcoded but relative to the available set)?

@Sudarshan-21 could you please give it a try and share a screenshot? I like the idea, this is probably the best compromise to have no distortion but still consistent card height

> In either case (distortion or not), having cards at the same height would be my preferred path going forward.

Thank you!

Sudarshan-21 commented:
> If it is simple, please do so and share a screenshot. I don't think it is easy to understand your concern (which might be right) without a screenshot, at least I find it a bit vague

@benoit74 The following is the UI after the suggested improvisation 

![Screenshot from 2024-04-08 19-19-53](https://github.com/openzim/kolibri/assets/137636943/ebe3530b-637e-4e69-88d0-9e48b56b7ae6)

benoit74 commented:
@Popolechien does your thumb up means the last proposition is your preferred solution?

I like it indeed, I would just ensure there is some space between the text and the "Open" button in all circumstances (e.g. "Caperucita Roja" is a bit too close to "Open" button)

Popolechien commented:
Yep, looks good to me (and agree on the padding requirement)

Sudarshan-21 commented:
@benoit74 all of the required changes are made, you can review this PR

![Screenshot from 2024-04-09 12-07-55](https://github.com/openzim/kolibri/assets/137636943/20709663-118f-4c2c-91b8-c2bcbe5a1c4d)

benoit74 reviewed (changes requested):
Your PR is not complete, you've made modifications only to the "default" cards, but not to "topic" and "more" cards, which have to be adapted as well to have consistent styling on all cards.

See conversations for additional details

Sudarshan-21 commented on zimui/src/components/TopicHome.vue:
```diff
@@ -191,15 +191,16 @@ const goToPreviousPage = () => {
         <div
           v-for="(content, contentIndex) in getNonTopicSections(topic.sections)"
           :key="contentIndex"
-          class="col-sm-6 col-md-6 col-lg-3 pt-3 pb-3"
```
Those paddings were within the card, having voids inside the div is unnecessary and the margin would serve the purpose of having space between cards.

Sudarshan-21 commented on zimui/src/components/TopicHome.vue:
```diff
@@ -191,15 +191,16 @@ const goToPreviousPage = () => {
       </div>
       <div class="row">
-        <footer class="pt-2 pb-3">
+        <footer class="mt-2 mb-1">
```
Having padding in the component is not needed. we just need to have some space between different elements which can be achieved by margin(which is a preferred way).

Sudarshan-21 commented on zimui/src/components/TopicHome.vue:
```diff
@@ -191,15 +191,16 @@ const goToPreviousPage = () => {
         >
           <TopicCard
             :data="transformTopicSectionOrSubSectionToCardData(content)"
+            class="w-100"
```
https://github.com/openzim/kolibri/issues/117#issuecomment-2025031828
the screenshot in this comment shows a card with reduced width, in order to fix the width the w-100 class is added.

benoit74 replied:
Got it, thank you

Sudarshan-21 commented on zimui/src/components/TopicCard.vue:
```diff
@@ -101,12 +101,18 @@ defineProps({
   transform: translateY(-50%);
 }
 
+.card-body{
```
This padding ensures the space between the title and the open button.

benoit74 replied:
Got it, thank you

Sudarshan-21 commented on zimui/src/components/TopicCard.vue:
```diff
@@ -101,12 +101,18 @@ defineProps({
+
 .card-img-top {
   border-radius: 1rem 1rem 0 0;
+  width: 100%; 
```
This is to ensure the image is not overflowing horizontally(I encountered that the last card image on the `minimal test` channel was horizontally overflowing)

benoit74 replied:
Got it, thank you

Sudarshan-21 commented on zimui/src/components/TopicCard.vue:
```diff
@@ -64,8 +64,8 @@ defineProps({
-  <div v-else class="card">
-    <a class="text-decoration-none text-reset" :href="`./files/${data.slug}`">
+  <div v-else class="card h-100">
+    <a class="text-decoration-none text-reset h-100 " :href="`./files/${data.slug}`">
```
`h-100` removed from `<a>`
![Screenshot from 2024-04-09 14-38-47](https://github.com/openzim/kolibri/assets/137636943/6996f733-724d-42e3-85d1-da936677c147)

`h-100` removed from <div>
![Screenshot from 2024-04-09 14-39-06](https://github.com/openzim/kolibri/assets/137636943/1b867463-8565-4f46-a34f-6493d3926230)

removing any of them results in shrinked card size.

Sudarshan-21 commented:
> Your PR is not complete, you've made modifications only to the "default" cards, but not to "topic" and "more" cards, which have to be adapted as well to have consistent styling on all cards.

@benoit74 there were no styles given to the `topic` and `more` classes, so do we want all the styles of `card` classes to be applied to these two classes?

benoit74 commented on zimui/src/components/TopicCard.vue:
```diff
@@ -101,12 +101,18 @@ defineProps({
   background-color: #091415;
   border-radius: 1rem;
   color: white;
+  overflow: hidden;
```
Got it, thank you

benoit74 commented:
> @benoit74 there were no styles given to the topic and more classes, so do we want all the styles of card classes to be applied to these two classes?

I'm not speaking about the `topic` and `more` classes, I'm speaking about the Vue/HTML code. Since for instance the "main" `div` all have at least the `card` class they inherit most of your changes but what is missing is other changes like the additional `h-100` and `open-badge` classes (btw, it shows that `open-badge` class should be renamed to something more generic since we want to apply this class to `topic` and `more` badges as well). Is it clearer?

Sudarshan-21 commented:
>all have at least the `card` class they inherit most of your changes but what is missing is other changes like the additional `h-100` 

I have moved the h-100 style to the card class itself. so this style will be applied whenever the card class is present.

>btw, it shows that `open-badge` class should be renamed to something more generic since we want to apply this class to `topic` and `more` badges as well)

I have renamed it, please check.
Is there anything more @benoit74?

benoit74 reviewed (changes requested):
`badge-wrapper` class is still missing for "MORE" and "EXPLORE" buttons

Sudarshan-21 commented:
@benoit74 I have made the required changes, let me know if there are more things to be taken care of.

benoit74 reviewed (approved):
LGTM, works well when the 3 kind of cards are present now:

![image](https://github.com/openzim/kolibri/assets/7102089/05c51f31-7098-4adf-8944-9c4a964fb114)

Thank you!

benoit74 commented:
I just squashed all commits into one and enhanced the commit message
